crates/hotshot-state-prover/src/circuit.rs

```diff
@@ -0,0 +1,219 @@
+
+        let threshold = u256_to_field::<F>(threshold);
+        let threshold_var = circuit.create_public_variable(threshold)?;
+
+        // TODO(Chengyu): put in the hotshot state
+        let public_inputs = vec![threshold];
+
+        // Checking whether the accumulated weight exceeds the quorum threshold
+        // We assume that NUM_ENTRIES is always a multiple of 2
+        let signed_amount_var = (0..STAKE_TABLE_CAPACITY / 2)
+            .map(|i| {
+                circuit.mul_add(
+                    &[
+                        stake_table_var[2 * i].stake_amount,
+                        signer_bit_vec_var[2 * i].0,
+                        stake_table_var[2 * i + 1].stake_amount,
+                        signer_bit_vec_var[2 * i + 1].0,
+                    ],
+                    &[F::one(), F::one()],
+                )
+            })
+            .collect::<Result<Vec<_>, CircuitError>>()?;
+        let acc_amount_var = circuit.sum(&signed_amount_var)?;
+        circuit.enforce_leq(threshold_var, acc_amount_var)?;
+
+        // circuit.mul_add(wires_in, q_muls)
+        circuit.finalize_for_arithmetization()?;
+        Ok((circuit, public_inputs))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{HotShotState, StateUpdateBuilder};
+    use ark_ed_on_bn254::EdwardsConfig as Config;
+    use ethereum_types::U256;
+    use hotshot_stake_table::vec_based::StakeTable;
+    use hotshot_types::traits::stake_table::StakeTableScheme;
+    use jf_primitives::signatures::{
+        bls_over_bn254::{BLSOverBN254CurveSignatureScheme, VerKey as BLSVerKey},
+        SchnorrSignatureScheme, SignatureScheme,
+    };
+    use jf_relation::Circuit;
+
+    type F = ark_ed_on_bn254::Fq;
+    type SchnorrVerKey = jf_primitives::signatures::schnorr::VerKey<Config>;
+
+    fn key_pairs_for_testing() -> Vec<(BLSVerKey, SchnorrVerKey)> {
+        let mut prng = jf_utils::test_rng();
+        (0..10)
+            .map(|_| {
+                (
+                    BLSOverBN254CurveSignatureScheme::key_gen(&(), &mut prng)
+                        .unwrap()
+                        .1,
+                    SchnorrSignatureScheme::key_gen(&(), &mut prng).unwrap().1,
+                )
+            })
+            .collect::<Vec<_>>()
+    }
+
+    fn stake_table_for_testing(
+        keys: &[(BLSVerKey, SchnorrVerKey)],
+    ) -> StakeTable<BLSVerKey, SchnorrVerKey, F> {
+        let mut st = StakeTable::<BLSVerKey, SchnorrVerKey, F>::new();
+        // Registering keys
+        keys.iter()
+            .for_each(|key| st.register(key.0, U256::from(100), key.1.clone()).unwrap());
+        // Freeze the stake table
+        st.advance();
+        st.advance();
+        st
+    }
+
+    #[test]
+    fn test_circuit_building() {
+        let keys = key_pairs_for_testing();
+        let st = stake_table_for_testing(&keys);
+
+        let bit_vec_6 = [
+            true, true, true, true, true, true, false, false, false, false,
```

Comment:
Maybe add a `false` in the consecutive `true` interval and add a `true` in the consecutive `false` interval.

Reply:
Done in de3a697